### ISSUE TITLE
Could you please compare SENet with a new neural net architecture called "EfficientNet" on KataGo?

### ISSUE THREAD

**Issue by poptangtwe** [lightvector/KataGo#42]
EfficientNet
https://ai.googleblog.com/2019/05/efficientnet-improving-accuracy-and.html
https://github.com/tensorflow/tpu/tree/master/models/official/efficientnet

**Comment by OmnipotentEntity** [comment]
I had to double check to make sure, but this isn't a type of neural network architecture.  It's a method for selecting efficient hyperparameters for a given neural network architecture.

In other words, there is nothing preventing SE (or GCNet for that matter) AND EfficientNet from being used at the same time.

However, that being said, image recognition tasks are much simpler conceptually than Go.  Image recognition tasks are trained on labeled data sets and tested on a known test set.  In Go training target is constantly changing as the net learns how to play.  I haven't read nearly enough of this paper to know if that is a problem to for EfficientNet approach though.

**Comment by isty2e** [comment]
Trying out AutoML before applying EffecientNet seems reasonable here, though experiments with auxiliary targets/features and other RL-accelerating tricks have higher priorities.

**Comment by alreadydone** [comment]
I would rather recommend that @lightvector try the method presented in the paper [*Large Memory Layers with Product Keys*](https://arxiv.org/abs/1907.05242) (see also [Lample](https://twitter.com/GuillaumeLample/status/1149646895377076224) and [LeCun](https://twitter.com/ylecun/status/1149678009617670144)'s tweets). The paper deals with transformers for NLP tasks; a memory-augmented 12-layer transformer outperformed a baseline 24-layer transformer. Increasing model capacity without slowing down inference is very attractive, since speed is crucial for game AI but 15b models are underfit compared to 40b models. The human approach to games involves not only development of intuition but also accumulation of knowledge in memory (or other storage media, like books); memory layers enable NNs to accumulate knowledge within itself. Increased capacity potentially allows the NNs to remember rare shapes and refutations to abandoned josekis, and to deal with liberties and ladders better (not so relevant for KataGo); as [McAllester pointed out](https://machinethoughts.wordpress.com/2019/06/25/the-inevitability-of-vector-quantization-in-deep-architectures/), such discretization/quantization/hard/sparse attention/dynamic routing mechanism employed in the key-value memory layers potentially reduces forgetting, so randomness in the opening for the sake of diversity in training data may not be necessary. Board games also has a more discrete flavor than images and involves more reasoning which makes it closer to visual question answering (VQA) tasks which have employed memory-augmented networks.

Here is my ready-for-action summary of the product-key memory method introduced in the paper, adapted to CNNs according to my understanding:

+ replace two (in the paper the 6th and 12th layers are replaced in a 16-layer model) of the convolutional layers (from d channels to d channels) by the following: a convolutional layer from d channels to dq channels; after batch normalization, the dq-dimensional vector at each of the 361 intersections on the board is split into two dq/2-dimensional query vectors. The memory is indexed by two subkeys, and each subkey can take |C| (learnable) possible values. For each of the two query vectors, find k nearest subkeys, retrieve the memory contents (learnable d-dimensional vectors) corresponding to the nearest k product keys, and average them in a certain way.

In the paper, d = 1024 or 1600, dq = 512, |C| = 512, memory size = 512^2.

**Comment by lightvector** [comment]
If we're still interested in this, it doesn't only have to be me to try all the stuff! Too much of a bottleneck. :)

A way that anyone else could contribute here, even without having a lot GPUs, only needing one GPU, would be for example to do supervised learning on KataGo training data. Implement the desired method in Tensorflow side-by-side with the existing one, and show that the new one works a lot better getting up to strong levels on the existing data or achieving low loss. Then, with the method already working in Tensorflow, it would be pretty trivial to just drop the architecture in to a subsequent real self-play run.

**Comment by lightvector** [comment]
Possibly relevant in the sense that it affects the model code and training data format - KataGo has added some new targets to help learn Japanese rules and is also experimenting with dropping batch norm entirely. By the end of this year, at which point the end of year run has actually started for real and is a healthy way along (despite some current delays in finalizing implementation details) I'll of course push all the new model code, trained models, and data.

**Comment by TonyCongqianWang** [comment]
Hi,

I am wondering if anyone did any experiments  in this area?

From the efficient net paper there are two insights:

1. They claimed thet lineat bottlenecks perform better empirically than bottlenecks with nonlinear activation functions after the 1x1 convolution layers.

2. Inverted linear bottlenecks in combination with separable conv blocks are very efficient w.r.t. parameters and flops.


That could be translated to two easy experiments to try. The experiments I propose would involve taking a small katago model architecture and modifying the trunk:

1. Remove activations after 1x1 conv layers in the bottleneck blocks.

2. Replace all nested bottleneck blocks with inverted bottleneck blocks, so that the flop count is slightly lower.

Then train on the self play data with multiple random seeds and compare the loss progression of the baseline model with the two modified models.

If the modified models have faster and better convergence, it is a good indication that the architecture is superior.

If nobody has tried this so far I'd like to try this in the future
